### DISCUSSION TITLE
My findings on performance

### DISCUSSION THREAD

**Discussion by FlixMa** [pocketbase/pocketbase#662] [Q&A] [answered]
Hey all!
I recently discovered that implicit relation expansions in filters take a long time to evaluate.

The architecture of my collections is simple:
A user may vote (*votes*) for a recipe (*recipes*) which will be available for serving in a specific week (*weeks*).

To get an overview of all the votes in a specific week one may design the following query:
```js
// fetch votes
const voteList = await pocketbase.records.getFullList('votes', 200, {
    filter: `recipe.week = "${weekId}"`,
});
```

However, when the entries in the collections grow (for me that is ~4k votes and ~2k recipes across ~90 weeks), the aforementioned query will take several seconds to execute (in my case ~10.8 seconds).
I confirmed this unfortunate behaviour of my react.js frontend by issuing a raw http request using Postman.

Currently, I am using the following workaround:
```js
// fetch recipes
const recipes = await pocketbase.records.getFullList('recipes', 200, {
  filter: `week = "${weekId}"`,
});

// fetch votes
// NOTE: this is a trick to avoid implicit expansion of relations -> brings down fetching time from 10.8s to 25ms
const recipeIdFilter = recipes.map((recipe) => `recipe = "${recipe.id}"`).join(' || ');
const voteList = await pocketbase.records.getFullList('votes', 200, {
    filter: recipeIdFilter
});
```

By fetching the recipes of said week first (roughly 35 recipes per week) and therefore probably avoiding the time-consuming implicit expansion, I am able to reduce the duration to only 25ms.

Is this behaviour intentional?
Anyway, I thought it might be interesting for some and helps designing your queries :-)

**Comment by ganigeorgiev** [comment, marked as answer]
What version of PocketBase are you using?

There were significant improvements to the relations lookup query in v0.7.4 (you could check discussion https://github.com/pocketbase/pocketbase/discussions/519; my tests were with 2 collections with 10k records each).

You can download the latest release from https://github.com/pocketbase/pocketbase/releases.

**Reply by FlixMa** [reply]
Oh, I missed that entry. Somehow the relevant respons of #519 didn’t show up or I haven’t scrolled far enough. Yes, I am indeed on 0.7.3. Sorry for the inconvenience!

**Reply by FlixMa** [reply]
So I just upgraded to 0.7.6 and I'm happy to confirm, that this indeed fixed the slow queries.

However, it broke my code at another point, where I used the `~` operator. 
It was due to this multi-relation field api check: `members ~ @request.user.profile.id`
#567 gave me a head start resolving this issue, but didn't completely solve it.

Eventually I found out, that it is necessary to use the `.id` property on the relation field like so:
`members.id ~ @request.user.profile.id`

This seems not to be the case when dealing with single relation fields as they may be queried like `user = "xyz"` (where *user* is the field name of a certain collection linking to the user).

If you knew this, just ignore it. Otherwise, now you know :-)
And many thanks for the quick help!

**Reply by ganigeorgiev** [reply]
@FlixMa Yep, sorry about this. I should have documented it better.

Previously the rule was working as a side effect of how `@request.user.profile.id` was being evaluated inconsistently, which may have resulted to expressions like `members LIKE '%PROFILE_ID%'`, where now this is always `members LIKE profile.id`. The current approach is better because when combined with the `json_each` changes in v0.7.4 it practically allow us to implement `IN`-like queries for relations without the hacky `LIKE` workaround, or in other words you can replace: `members.id ~ @request.user.profile.id` with `members.id = @request.user.profile.id` and it should work just fine. There still a lot of room for improvements though.

**Reply by ganigeorgiev** [reply]
_Just to let you know that I've pushed in `master` a fallback handling for the contains operator when both operands are table columns (as in `members ~ @request.user.profile.id`) to minimize the breaking changes (it will be available in v0.7.7 sometime later this week). In any case, your approach with the `members.id =/~ ...` is the recommended way for relational lookups._
